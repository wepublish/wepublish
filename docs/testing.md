# Automated Testing

Automated testing is a crucial aspect of software development that helps to ensure the quality of a codebase by
automating the testing process.

## API Unit Testing

API unit testing is a type of automated testing that focuses on testing individual units of code that make up an API.

### Setting up the Testing Environment

To set up the testing environment, you need to ensure that you have the following entry in your `jest.config.ts` file: `globalSetup: '<rootDir>/setup-database.js'`. You can find an example of this in the `libs/membership/api/src` folder.

In the `setup-database.js` file, you should have code that looks like the following:

```typescript
import execa from 'execa'

export default async () => {
    const databaseUrl = 'postgresql://postgres@localhost:5432/wepublish_test?schema=public'
    process.env.DATABASE_URL = databaseUrl
    process.env.TZ = 'UTC'

    await execa(`npx`, ['prisma', 'migrate', 'reset', '--force'])
}
```
This code sets up a test database called `wepublish_test` and resets it each time you run a test.

### Running Tests
To run tests, you need to ensure that all your changes to the `schema.prisma` file are reflected in a migration. 
You can do this by running `prisma migrate dev --name name_of_my_migration`. Otherwise, your test database may 
rely on an incomplete schema.
Also, make sure, you have a database up and running. You may want run `docker-compose up database`

Once you have done this, you should be able to run a single test using the command `npx jest -t 'my test'`

### Writing Tests with Fabbrica
When it comes to writing tests, we highly recommend using [Fabbrica](https://github.com/Quramy/prisma-fabbrica).
With Fabbrica, you can easily generate database entries using the `.create()` method or model instances using
the `.build()` method, saving you time and effort.

For examples of how to use Fabbrica in your tests, check out the `subscription-flow.controller.spec.ts` file 
located at `libs/membership/api/src/lib/subscription-flow`. It's a great resource to help you get started with writing
tests that utilize Fabbrica.

### ! Do not forget adding tests to the pipeline !
In order to run your tests within the pipeline, you have to add an entry in the `package.json` in the root folder.
To test, if it works, run `nx test` from your cli. Your specs should now be tested.


## Snapshots
It might happen that your tests are failing. Try to update your snapshots
by running (eventually in the package in question) `npm run test -- -u`

For more information: [https://jestjs.io/docs/cli#--updatesnapshot](https://jestjs.io/docs/cli#--updatesnapshot)