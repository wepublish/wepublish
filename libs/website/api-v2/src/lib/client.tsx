import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
