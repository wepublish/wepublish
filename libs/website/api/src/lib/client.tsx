import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
