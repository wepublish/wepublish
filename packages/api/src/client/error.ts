export enum ErrorCode {
  TokenExpired = 'tokenExpired',
  InvalidToken = 'invalidToken',
  Unauthorized = 'unauthorized',
  InvalidCredentials = 'invalidCredentials'
}
