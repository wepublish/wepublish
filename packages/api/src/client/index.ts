export * from './types'
export * from './utility'
export * from './queryArticles'
