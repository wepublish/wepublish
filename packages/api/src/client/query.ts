import axios from 'axios'

const request = axios.create()

export class CancelError extends Error {}

export class CancelToken {
  public source = axios.CancelToken.source()

  cancel() {
    this.source.cancel()
  }
}

export interface RequestOptions {
  cancelToken?: CancelToken
}

export interface QueryOptions {
  query?: string
  variables?: {[key: string]: any}
}

export async function query(
  url: string,
  {query = '', variables}: QueryOptions,
  {cancelToken}: RequestOptions = {}
): Promise<any> {
  try {
    const result = await request.post(
      url,
      {
        query,
        variables
      },
      {
        cancelToken: cancelToken && cancelToken.source.token
      }
    )

    return result.data.data
  } catch (err) {
    if (axios.isCancel(err)) throw new CancelError()
    throw err
  }
}
