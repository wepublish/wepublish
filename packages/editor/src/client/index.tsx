import 'core-js/stable'
import 'regenerator-runtime/runtime'

import React from 'react'
import ReactDOM from 'react-dom'
import {ApolloProvider, ApolloClient, ApolloLink, InMemoryCache} from '@apollo/client'
import {createUploadLink} from 'apollo-upload-client'
import {onError} from '@apollo/client/link/error'

import './i18n'

import {ElementID} from '../shared/elementID'
import {ClientSettings} from '../shared/types'
import {RouteProvider} from './route'
import {AuthProvider} from './authContext'
import {LocalStorageKey} from './utility'
import {TwitterProvider} from './blocks/embeds/twitter'
import {InstagramProvider} from './blocks/embeds/instagram'
import {FacebookProvider} from './blocks/embeds/facebook'
import {HotApp} from './app'

// See: https://www.apollographql.com/docs/react/data/fragments/#fragments-on-unions-and-interfaces
export async function fetchIntrospectionQueryResultData(url: string) {
  const response = await fetch(url, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify({
      variables: {},
      query: `
      {
        __schema {
          types {
            kind
            name
            possibleTypes {
              name
            }
          }
        }
      }
    `
    })
  })

  const result = await response.json()

  const possibleTypes: any = {}

  result.data.__schema.types.forEach((supertype: any) => {
    if (supertype.possibleTypes) {
      possibleTypes[supertype.name] = supertype.possibleTypes.map((subtype: any) => subtype.name)
    }
  })

  return possibleTypes
}

const onDOMContentLoaded = async () => {
  const {apiURL}: ClientSettings = JSON.parse(
    document.getElementById(ElementID.Settings)!.textContent!
  )

  const adminAPIURL = `${apiURL}/admin`

  const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(LocalStorageKey.SessionToken)
    const context = operation.getContext()

    operation.setContext({
      headers: {
        authorization: token ? `Bearer ${token}` : '',
        ...context.headers
      },
      ...context
    })

    return forward(operation)
  })

  const authErrorLink = onError(({graphQLErrors, /* networkError, */ operation, forward}) => {
    if (graphQLErrors) {
      graphQLErrors.forEach(({/* message, locations, path, */ extensions}) => {
        // console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`)
        if (extensions?.code === 'UNAUTHENTICATED') {
          window.location.pathname = '/logout'
        }
      })
      // if (networkError) console.log(`[Network error]: ${networkError}`)
    }
    return forward(operation)
  })

  const client = new ApolloClient({
    link: authLink.concat(authErrorLink).concat(createUploadLink({uri: adminAPIURL})),
    cache: new InMemoryCache({
      possibleTypes: await fetchIntrospectionQueryResultData(adminAPIURL)
    })
  })

  window.addEventListener('dragover', e => e.preventDefault())
  window.addEventListener('drop', e => e.preventDefault())

  ReactDOM.render(
    <ApolloProvider client={client}>
      <AuthProvider>
        <RouteProvider>
          <FacebookProvider sdkLanguage={'en_US'}>
            <InstagramProvider>
              <TwitterProvider>
                <HotApp />
              </TwitterProvider>
            </InstagramProvider>
          </FacebookProvider>
        </RouteProvider>
      </AuthProvider>
    </ApolloProvider>,
    document.getElementById(ElementID.ReactRoot)
  )
}

if (document.readyState !== 'loading') {
  onDOMContentLoaded()
} else {
  document.addEventListener('DOMContentLoaded', onDOMContentLoaded)
}
