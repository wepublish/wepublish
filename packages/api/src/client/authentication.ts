import {query, RequestOptions} from './query'

export async function authenticateWithCredentials(
  url: string,
  email: string,
  password: string,
  opts?: RequestOptions
): Promise<{
  user: {email: string}
  refreshToken: string
  accessToken: string
  refreshTokenExpiresIn: number
  accessTokenExpiresIn: number
}> {
  const {authenticateWithCredentials: response} = await query(
    url,
    {
      query: `
    mutation authenticateWithCredentials($email: String!, $password: String!) {
      authenticateWithCredentials(email: $email, password: $password) {
        user {email},
        refreshToken,
        accessToken,
        refreshTokenExpiresIn,
        accessTokenExpiresIn
      }
    }
  `,
      variables: {
        email,
        password
      }
    },
    opts
  )
  return response
}
