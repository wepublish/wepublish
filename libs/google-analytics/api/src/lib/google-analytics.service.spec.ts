import {Test, TestingModule} from '@nestjs/testing'
import {PrismaClient} from '@prisma/client'
import {
  GA_CLIENT_OPTIONS,
  GoogleAnalyticsConfig,
  GoogleAnalyticsService
} from './google-analytics.service'

const runReportSpy = jest.fn()

jest.mock('@google-analytics/data', () => ({
  BetaAnalyticsDataClient: jest.fn().mockImplementation(() => ({
    runReport: runReportSpy
  }))
}))

describe('GoogleAnalyticsService', () => {
  let config: GoogleAnalyticsConfig
  let service: GoogleAnalyticsService
  let prismaMock: {
    article: {[method in keyof PrismaClient['article']]?: jest.Mock}
  }

  beforeAll(() => {
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01'))
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(async () => {
    config = {
      credentials: {},
      articlePrefix: '/a/',
      property: '1234'
    }

    prismaMock = {
      article: {
        count: jest.fn(),
        findMany: jest.fn(),
        findUnique: jest.fn(),
        delete: jest.fn(),
        create: jest.fn(),
        update: jest.fn()
      }
    }

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        GoogleAnalyticsService,
        {provide: PrismaClient, useValue: prismaMock},
        {
          provide: GA_CLIENT_OPTIONS,
          useValue: config
        }
      ]
    }).compile()

    service = module.get<GoogleAnalyticsService>(GoogleAnalyticsService)
  })

  it('should return an empty array when property is not set', async () => {
    config.property = undefined
    const result = await service.getMostViewedArticles({})

    expect(result).toHaveLength(0)
    expect(runReportSpy).not.toHaveBeenCalled()
  })

  it('should return  an empty array when credentials is not set', async () => {
    config.credentials = undefined
    const result = await service.getMostViewedArticles({})

    expect(result).toHaveLength(0)
    expect(runReportSpy).not.toHaveBeenCalled()
  })

  it('should get articles by popularity', async () => {
    runReportSpy.mockReturnValue(
      Promise.resolve([
        {
          rows: [
            {
              dimensionValues: [{value: 'foobar'}],
              metricValues: [{value: '123'}]
            },
            {
              dimensionValues: [{value: 'barfoo'}],
              metricValues: [{value: '1234'}]
            },
            {
              dimensionValues: [{value: 'bazfoo'}],
              metricValues: [{value: '123456'}]
            },
            {
              dimensionValues: [{value: 'foobaz'}],
              metricValues: [{value: '12345'}]
            }
          ]
        }
      ])
    )
    prismaMock.article.findMany?.mockReturnValue([
      {published: {slug: 'foobar'}},
      {published: {slug: 'barfoo'}},
      {published: {slug: 'bazfoo'}},
      {published: {slug: 'foobaz'}}
    ])

    const result = await service.getMostViewedArticles({})

    expect(result).toMatchSnapshot()
    expect(prismaMock.article.findMany?.mock.calls[0]).toMatchSnapshot()
  })
})
