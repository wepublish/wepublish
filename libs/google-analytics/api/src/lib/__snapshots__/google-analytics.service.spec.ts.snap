// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleAnalyticsService should get articles by popularity 1`] = `
[
  {
    "published": {
      "slug": "foobaz",
    },
  },
  {
    "published": {
      "slug": "bazfoo",
    },
  },
  {
    "published": {
      "slug": "barfoo",
    },
  },
  {
    "published": {
      "slug": "foobar",
    },
  },
]
`;

exports[`GoogleAnalyticsService should get articles by popularity 2`] = `
[
  {
    "include": {
      "published": {
        "select": {
          "slug": true,
        },
      },
    },
    "skip": 0,
    "take": 10,
    "where": {
      "published": {
        "slug": {
          "in": [],
          "mode": "insensitive",
        },
      },
    },
  },
]
`;
